src/module.c

```diff
@@ -144,74 +144,77 @@ struct ValkeyModuleBlockedClient;
+    struct AutoMemEntry *amqueue;                     /* Auto memory queue of objects to free. */
+    int amqueue_len;                                  /* Number of slots in amqueue. */
+    int amqueue_used;                                 /* Number of used slots in amqueue. */
+    int flags;                                        /* VALKEYMODULE_CTX_... flags. */
+    void **postponed_arrays;                          /* To set with VM_ReplySetArrayLength(). */
+    int postponed_arrays_count;                       /* Number of entries in postponed_arrays. */
+    void *blocked_privdata;                           /* Privdata set when unblocking a client. */
+    ValkeyModuleString *blocked_ready_key;            /* Key ready when the reply callback
+                                                        gets called for clients blocked
+                                                        on keys. */
+
+    /* Used if there is the VALKEYMODULE_CTX_KEYS_POS_REQUEST or
      * VALKEYMODULE_CTX_CHANNEL_POS_REQUEST flag set. */
     getKeysResult *keys_result;
 
     struct ValkeyModulePoolAllocBlock *pa_head;
     long long next_yield_time;
 
-    const struct ValkeyModuleUser *user;  /* ValkeyModuleUser commands executed via
-                                            VM_Call should be executed as, if set */
+    const struct ValkeyModuleUser *user; /* ValkeyModuleUser commands executed via
+                                           VM_Call should be executed as, if set */
 };
 typedef struct ValkeyModuleCtx ValkeyModuleCtx;
 
 #define VALKEYMODULE_CTX_NONE (0)
-#define VALKEYMODULE_CTX_AUTO_MEMORY (1<<0)
-#define VALKEYMODULE_CTX_KEYS_POS_REQUEST (1<<1)
-#define VALKEYMODULE_CTX_BLOCKED_REPLY (1<<2)
-#define VALKEYMODULE_CTX_BLOCKED_TIMEOUT (1<<3)
-#define VALKEYMODULE_CTX_THREAD_SAFE (1<<4)
-#define VALKEYMODULE_CTX_BLOCKED_DISCONNECTED (1<<5)
-#define VALKEYMODULE_CTX_TEMP_CLIENT (1<<6) /* Return client object to the pool
-                                              when the context is destroyed */
-#define VALKEYMODULE_CTX_NEW_CLIENT (1<<7)  /* Free client object when the
-                                              context is destroyed */
-#define VALKEYMODULE_CTX_CHANNELS_POS_REQUEST (1<<8)
-#define VALKEYMODULE_CTX_COMMAND (1<<9) /* Context created to serve a command from call() or AOF (which calls cmd->proc directly) */
+#define VALKEYMODULE_CTX_AUTO_MEMORY (1 << 0)
+#define VALKEYMODULE_CTX_KEYS_POS_REQUEST (1 << 1)
+#define VALKEYMODULE_CTX_BLOCKED_REPLY (1 << 2)
+#define VALKEYMODULE_CTX_BLOCKED_TIMEOUT (1 << 3)
+#define VALKEYMODULE_CTX_THREAD_SAFE (1 << 4)
+#define VALKEYMODULE_CTX_BLOCKED_DISCONNECTED (1 << 5)
+#define VALKEYMODULE_CTX_TEMP_CLIENT                                                                                   \
+    (1 << 6) /* Return client object to the pool                                                                       \
+               when the context is destroyed */
```

Comment:
Yeah, that's also a good example of why trailing line comments are bad in practice.

It's much cleaner to do:

```c
/* comment for the next thing */
#define VALKEYMODULE_CTX_TEMP_CLIENT (1 << 6)
```

It would require manually fixing all those broken end-of-line comments though.

Reply:
I'm surprised this even compiles.  Don't you need to escape the new-line after `...CLIENT` ?

Reply:
>  Don't you need to escape the new-line after ...CLIENT ?

if you scroll far to the right, the `\` is right-aligned at full line spec width (the github preview just has it out of scope with no wide scroll indicator)

> experiments in code-understanding concluded that people understand trailing-line comments more easily.

seems illogical? vertical parsing is clearly superior.

also this example is weird because it's a macro. including a comment in-line with a macro means the comment is also being expanded into the macro replacement locations (which works, but is also unnecessary).

Reply:
If you have to scroll all the way to the maximum line width in order to find something, then I think there is something wrong.  If we decide to apply this kind of formatting, I would ask that the `\` be placed very close after `...CLIENT` and `...pool`.

> seems illogical? vertical parsing is clearly superior.

I have seen this question about placement of comments lead to many religious arguments.  After too many of them
I became involved in a project to build a web-site of code-comprehension tools and exercises for the University of Toronto.  This was many years ago, and now, sadly, the web-site does not exist.  I distinctly recall that while doing research for that project I read a paper in which they performed experiments with multiple programmers reading code and answering questions about the code.  I was impressed by the rigor of the methodology.  Again, sadly, I am only able to report the results as I remember them.  The reason I remember them is because the paper provided an empirical answer to this question which had vexed me for so long.

In spite of all this, I truly believe that there is no one right answer here.  This is about communicating to humans.  I think the correct answer is "it depends".

Reply:
> If you have to scroll all the way to the maximum line width in order to find something

It's kinda an artifact of somebody setting the line width unnaturally long too. Normally you just set 80 chars and everything looks fine (and normally we don't develop in github quoted preview blocks etc).

As for end of line comments, here we go using standard tools:
<img width="1018" alt="Screenshot 2024-04-17 at 12 55 13" src="https://github.com/valkey-io/valkey/assets/15594/f2b0f7ad-2532-44c8-9217-42411498da97">
<img width="1018" alt="Screenshot 2024-04-17 at 12 55 04" src="https://github.com/valkey-io/valkey/assets/15594/98020529-acaa-47ee-a037-c16723a6f1af">

Reply:
Sorry, I don't believe anyone would write end-of-line comments like that.  That looks like a straw-man argument to me.

Suppose the original code looks like this:
~~~
#define VALKEYMODULE_CTX_TEMP_CLIENT (1 << 6) /* Return client object to the pool  when the context is destroyed */
~~~

1) This is already terrible code!

2) Yes, this would look MUCH better:
~~~
/* Return client object to the pool  when the context is destroyed */
#define VALKEYMODULE_CTX_TEMP_CLIENT (1 << 6)
~~~

3) If it is going to be auto-formatted, then this would be horrible but tolerable for me:
~~~
#define VALKEYMODULE_CTX_TEMP_CLIENT \
    (1 << 6) /* Return client object to the pool \
                    when the context is destroyed */
~~~
But I would much prefer this horrible result:
~~~
#define VALKEYMODULE_CTX_TEMP_CLIENT               \
    (1 << 6) /* Return client object to the pool   \
                    when the context is destroyed */
~~~

4) This still feels like a straw-man to me.  Even a modestly smarter formatter should give
~~~
12345678901234567890123456789012345678901234567890123456789012345678901234567890

#define VALKEYMODULE_CTX_TEMP_CLIENT (1 << 6) /* Return client object to the \
                                                 pool when the context is    \
                                                 destroyed */
~~~
Note: the first line is a ruler.  It is exactly 80 characters long.

5) Is (4) more readable than (2)?  That probably depends on both the context and the reader.  Consider this context:
~~~
12345678901234567890123456789012345678901234567890123456789012345678901234567890

#define VALKEYMODULE_CTX_JUNK        (1 << 5) /* Do something with junk objects. */
#define VALKEYMODULE_CTX_TEMP_CLIENT (1 << 6) /* Return client object to the pool when the context is destroyed */
#define VALKEYMODULE_CTX_TRASH       (1 << 7) /* Don't forget to take out the trash every Thursday night. */
~~~

It violates the 80 character rule.  In this case, I would say violating the rule is justified.

6) Would you prefer this?
~~~
/* Do something with junk objects. */
#define VALKEYMODULE_CTX_JUNK (1 << 5)

/* Return client object to the pool when the context is destroyed */
#define VALKEYMODULE_CTX_TEMP_CLIENT (1 << 6)

/* Don't forget to take out the trash every Thursday night. */
#define VALKEYMODULE_CTX_TRASH (1 << 7)
~~~

7) In summary,
* Communication is **HARD**.
* Guidelines are good, absolute rules are bad.
* Context is critical.
* Different people respond in different ways to the same thing.

Reply:
Extremely well thought through 👍 

Yeah, your 5 vs 6 example is good because it depends on the use case. It's nice seeing the vertical alignment to compare values, but the per-line comments overflow the viewport; while the 6 is easier to read the comments but takes an extra second to very 5->6->7 are in order.

Overall, going for nice machine style enforcing clean maintainable readability is always a win over "big programmer brain decide style decision per-line" (at least when working in groups and on projects that can live for decades, for single projects just do whatever, which is why it's always nice to have single projects too!).

overall these are just feedback ideas for whoever makes actual decisions around here 🦅 

## values!

This is also a bit of a weird/contrived example because technically these things should be enums and not defines anyway 🙃 

oh glob i just looked and the enums in the code are all weird too. They are almost all anonymous `typedef enum { a, b, c } name;` so they won't show their fully type details in debuggers versus proper usage of `typedef enum name { a, b, c } name;` wheeeee:


```haskell
src/$ rg "typedef enum" --stats

37 matches
37 matched lines
18 files contained matches
```

## alignment

as for right-aligned continues, here's some live code found in the wild:

```c
#define INSERT_SEP \   
    if (g->state[g->depth] == yajl_gen_map_key ||               \
        g->state[g->depth] == yajl_gen_in_array) {              \
        g->print(g->ctx, ",", 1);                               \
        if ((g->flags & yajl_gen_beautify)) g->print(g->ctx, "\n", 1);               \
    } else if (g->state[g->depth] == yajl_gen_map_val) {        \
        g->print(g->ctx, ":", 1);                               \
        if ((g->flags & yajl_gen_beautify)) g->print(g->ctx, " ", 1);                \
   }

#define INSERT_WHITESPACE                                               \
    if ((g->flags & yajl_gen_beautify)) {                                                    \
        if (g->state[g->depth] != yajl_gen_map_val) {                   \
            unsigned int _i;                                            \
            for (_i=0;_i<g->depth;_i++)                                 \
                g->print(g->ctx,                                        \
                         g->indentString,                               \
                         (unsigned int)strlen(g->indentString));        \
        }                                                               \
    }                               
                                     
#define ENSURE_NOT_KEY \                         
    if (g->state[g->depth] == yajl_gen_map_key ||       \
        g->state[g->depth] == yajl_gen_map_start)  {    \
        return yajl_gen_keys_must_be_strings;           \
    }                                                   \ 

/* initialize a bytestack */
#define yajl_bs_init(obs, _yaf) {               \
        (obs).stack = NULL;                     \
        (obs).size = 0;                         \
        (obs).used = 0;                         \
        (obs).yaf = (_yaf);                     \
    }                                           \


/* initialize a bytestack */
#define yajl_bs_free(obs)                 \
    if ((obs).stack) (obs).yaf->free((obs).yaf->ctx, (obs).stack);

#define yajl_bs_current(obs)               \
    (assert((obs).used > 0), (obs).stack[(obs).used - 1])

#define yajl_bs_push(obs, byte) {                       \
    if (((obs).size - (obs).used) == 0) {               \
        (obs).size += YAJL_BS_INC;                      \
        (obs).stack = (obs).yaf->realloc((obs).yaf->ctx,\
                                         (void *) (obs).stack, (obs).size);\
    }                                                   \
    (obs).stack[((obs).used)++] = (byte);               \
}

```

and here's a better auto-format cleaned-up version (still not perfect; executable code statements in defines should be in a `do { } while (0);` loop, etc, but that's behavior/correctness issues instead of formatting):

```c
#define INSERT_SEP                                                             \
    if (g->state[g->depth] == yajl_gen_map_key ||                              \
        g->state[g->depth] == yajl_gen_in_array) {                             \
        g->print(g->ctx, ",", 1);                                              \
        if ((g->flags & yajl_gen_beautify))                                    \
            g->print(g->ctx, "\n", 1);                                         \
    } else if (g->state[g->depth] == yajl_gen_map_val) {                       \
        g->print(g->ctx, ":", 1);                                              \
        if ((g->flags & yajl_gen_beautify))                                    \
            g->print(g->ctx, " ", 1);                                          \
    }                    
                     
#define INSERT_WHITESPACE                                                      \
    if ((g->flags & yajl_gen_beautify)) {                                      \
        if (g->state[g->depth] != yajl_gen_map_val) {                          \
            unsigned int _i;                                                   \
            for (_i = 0; _i < g->depth; _i++)                                  \
                g->print(g->ctx, g->indentString,                              \
                         (unsigned int)strlen(g->indentString));               \
        }                                                                      \
    }         
                   
#define ENSURE_NOT_KEY                                                         \
    if (g->state[g->depth] == yajl_gen_map_key ||                              \
        g->state[g->depth] == yajl_gen_map_start) {                            \
        return yajl_gen_keys_must_be_strings;                                  \
    }                 

/* initialize a bytestack */
#define yajl_bs_init(obs, _yaf)                                                \
    {                                                                          \
        (obs).stack = NULL;                                                    \
        (obs).size = 0;                                                        \
        (obs).used = 0;                                                        \
        (obs).yaf = (_yaf);                                                    \
    }

/* initialize a bytestack */
#define yajl_bs_free(obs)                                                      \
    if ((obs).stack)                                                           \
        (obs).yaf->free((obs).yaf->ctx, (obs).stack);

#define yajl_bs_current(obs)                                                   \
    (assert((obs).used > 0), (obs).stack[(obs).used - 1])

#define yajl_bs_push(obs, byte)                                                \
    {                                                                          \
        if (((obs).size - (obs).used) == 0) {                                  \
            (obs).size += YAJL_BS_INC;                                         \
            (obs).stack = (obs).yaf->realloc((obs).yaf->ctx,                   \
                                             (void *)(obs).stack, (obs).size); \
        }                                                                      \
        (obs).stack[((obs).used)++] = (byte);                                  \
    }

```

Reply:
Speaking of
> big programmer brain decide style decision per-line
see https://en.wikipedia.org/wiki/Donald_Knuth and https://en.wikipedia.org/wiki/Literate_programming

Reply:
This is just an automation tool doing its magic. 

I would prefer *not* having trailing comments but I don't know if clang-format can handle that. 

clang-format is the best free tool that I know of so I am willing to work with its limitations in exchange of a consistent style (even if it is not my favorite). In other words, you can say my bar is low, just "consistency" :-)

Reply:
But we can decide the our line limit is longer than 80, which is a very old limitation dating back to CRT monitors and VI vs Emacs wars.
I can easily live with 120 or even 160 line limit.
To make a point my current screen is so wide (and curved) that most of this page is white space to either side and a narrow band of thread in the middle.